cmd/bitrot.go

```diff
@@ -129,26 +130,31 @@ func newBitrotReader(disk StorageAPI, volume, filePath string, algo BitrotAlgori
 		filePath:  filePath,
 		verifier:  &BitrotVerifier{algo, sum},
 		endOffset: endOffset,
-		buf:       nil,
 	}
 }
 
 // ReadChunk returns requested data.
 func (b *bitrotReader) ReadChunk(offset int64, length int64) ([]byte, error) {
-	if b.buf == nil {
-		b.buf = make([]byte, b.endOffset-offset)
-		if _, err := b.disk.ReadFile(b.volume, b.filePath, offset, b.buf, b.verifier); err != nil {
-			logger.LogIf(context.Background(), err)
-			return nil, err
-		}
+	var err error
+	var rc io.ReadCloser
+	if rc, err = b.disk.ReadFile(b.volume, b.filePath, offset, length, b.verifier); err != nil {
+		logger.LogIf(context.Background(), err)
+		return nil, err
 	}
-	if int64(len(b.buf)) < length {
-		logger.LogIf(context.Background(), errLessData)
+
+	var buf []byte
+	buf, err = ioutil.ReadAll(rc)
+	rc.Close()
```

Comment:
The code is inefficient, it does bitrot verification for each ReadChunk() call, we should do bitrot verirication only once. In posix.go, we should hold the buffer in memory instead of re-reading from disk for every ReadChunk() call

Reply:
The function should look like this:
```
// ReadChunk returns requested data.
func (b *bitrotReader) ReadChunk(offset int64, length int64) ([]byte, error) {
        var err error
        if b.reader == nil {
                if b.reader, err = b.disk.ReadFile(b.volume, b.filePath, offset, b.endOffset, b.verifier); err != nil {
                        logger.LogIf(context.Background(), err)
                        return nil, err
                }
        }

        buf := make([]byte, length)
        n, err := b.reader.Read(buf)
        if err != nil && err != io.EOF {
                return nil, err
        }

        if int64(n) != length {
                return nil, errLessData
        }

        return buf, nil
}
```

Reply:
As I used your unmodified change previously, unit tests fail with code.  That needs to be identified and fixed. So that I can take the change.

Reply:
Sure you can fix the test cases too. Note that you dont have to take my change as is, it was just some code pointers.

Reply:
Note that there is no problems with those tests.  The problem was the code snippet you pasted.  That's the reason I modified the code to make it work.  This means the assumption of "code inefficient" is wrong.

Reply:
@balamurugana I have sent a PR to your branch with the needed changes.

There is one more memory related improvement you can do.

You can notice that `ReadChunk()` creates a buffer for every `ReadChunk()` call.
What can be done is, you can create the buffer once in `newParallelReader()` 
```
// Reads in parallel from bitrotReaders.
type parallelReader struct {
        readers       []*bitrotReader
        dataBlocks    int
        offset        int64
        shardSize     int64
        shardFileSize int64
        blocks        [][]byte
}
```

i.e you can have this new field `blocks` which can be re-used

Reply:
I think this change is outside of the scope of this PR.  Its not a good idea to do many changes in reader support.

Reply:
@balamurugana best place to allocate the reusable memory is at the caller of ` func (s ErasureStorage) ReadFile()